### PR TITLE
Only process or load rows if not empty

### LINKED ISSUE #24561: SANS Load has error if empty rows present
**Original reporter:** nobody

### Expected behavior
load to ignore empty rows in the table

### Actual behavior
SANS will attempt to load using data (not) present in an empty row, which will obviously fail.

### Steps to reproduce the behavior
Load any data in the SANS v2 GUI, then insert a row and click load

### Platforms affected
All

### PR DESCRIPTION
**Pull request by TTitcombe**
**Description of work.**
Adds an additional check before attempting to load or process rows so that only non-empty table rows are included.

**Report to:** nobody

**To test:**
[PRdata_no_runs.zip](https://github.com/mantidproject/mantid/files/2812162/PRdata_no_runs.zip)
1. Interfaces -> SANS -> SANS v2
2. Load the sans2d user and batch file in the attached data
3. Click the "insert row" button to add any number of empty rows
4. Click load, process all, or process selected (highlight any rows to be included in processing for this one)
5. No errors or warnings should occur

Fixes #24561 
---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by robertapplin** [comment]
@TTitcombe I get this warning when clicking `Process All` or when clicking `Process Selected` when the filled row is selected. Is this expected?
image

File: scripts/SANS/sans/gui_logic/models/table_model.py [hunk 1]
```diff
@@ -128,6 +128,14 @@ def create_empty_row():
         row = [''] * 16
         return TableIndexModel(*row)
 
+    def get_non_empty_rows(self, rows):
+        non_empty_rows = []
+        for index in rows:
+            row = self.get_table_entry(index)
+            if not row.is_empty():
+                non_empty_rows.append(index)
+        return non_empty_rows
```
Comment (robertapplin) [inline scripts/SANS/sans/gui_logic/models/table_model.py]: Could you use a list comprehension here?
`non_empty_rows = [index for index in rows if not self.get_table_entry(index).is_empty()]`

**Review by robertapplin — CHANGES_REQUESTED** [review]
The changes look good and the tests have passed. Just an optional comment about maybe using a list comprehension to possibly improve readability?

File: scripts/SANS/sans/gui_logic/models/table_model.py (hunk 1 above)
Comment (TTitcombe) [reply scripts/SANS/sans/gui_logic/models/table_model.py]: I agree this is unnecessarily terse. I've changed it to use filter (filter is generally faster than list comprehensions)

**Comment by TTitcombe** [comment]
@robertapplin I think it may be that it can't find the runs. Do you have ISIS data archive searching in manage directories turned off?

**Review by robertapplin — DISMISSED** [review]

**Review by robertapplin — APPROVED** [review]

**Pull request merged**